File: user-service/src/main/java/rs/raf/user_service/controller/EmployeeController.java
package rs.raf.user_service.controller;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.boot.context.properties.bind.validation.ValidationErrors;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.validation.BindingResult;
import org.springframework.validation.ObjectError;
import org.springframework.web.bind.annotation.*;
import rs.raf.user_service.dto.*;
import rs.raf.user_service.exceptions.EmailAlreadyExistsException;
import rs.raf.user_service.exceptions.UserAlreadyExistsException;
import rs.raf.user_service.service.EmployeeService;

import javax.persistence.EntityNotFoundException;
import javax.validation.Valid;
import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/admin/employees")
@Tag(name = "Employee Management", description = "API for managing employees")
public class EmployeeController {

    private final EmployeeService employeeService;

    public EmployeeController(EmployeeService employeeService) {
        this.employeeService = employeeService;
    }

    @PreAuthorize("hasAuthority('admin')")
    @Operation(summary = "Get employee by ID", description = "Returns an employee based on the provided ID")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Successfully retrieved employee"),
            @ApiResponse(responseCode = "404", description = "Employee not found")
    })
    @GetMapping("/{id}")
    public ResponseEntity<?> getEmployeeById(@PathVariable Long id) {
        try {
            return ResponseEntity.ok().body(employeeService.findById(id));
        }
        catch(EntityNotFoundException e) {
            return ResponseEntity.notFound().build();
        }

    }

    @PreAuthorize("hasAuthority('admin')")
    @Operation(summary = "Get all employees", description = "Returns a paginated list of employees with optional filters")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Successfully retrieved employee list")
    })
    @GetMapping
    public ResponseEntity<Page<EmployeeDto>> getAllEmployees(
            @RequestParam(required = false) String position,
            @RequestParam(required = false) String department,
            @RequestParam(required = false) Boolean active,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {

        Pageable pageable = PageRequest.of(page, size);
        return ResponseEntity.ok(employeeService.findAll(position, department, active, pageable));
    }

    @PreAuthorize("hasAuthority('admin')")
    @Operation(summary = "Delete an employee", description = "Deletes an employee by their ID.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Employee deleted successfully"),
            @ApiResponse(responseCode = "404", description = "Employee not found")
    })
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteEmployee(
            @Parameter(description = "Employee ID", required = true, example = "1")
            @PathVariable Long id) {
        try {
            employeeService.deleteEmployee(id);
            return ResponseEntity.ok().build();
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    @PreAuthorize("hasAuthority('admin')")
    @Operation(summary = "Deactivate an employee", description = "Deactivates an employee by their ID.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Employee deactivated successfully"),
            @ApiResponse(responseCode = "404", description = "Employee not found")
    })
    @PatchMapping("/{id}/deactivate")
    public ResponseEntity<Void> deactivateEmployee(
            @Parameter(description = "Employee ID", required = true, example = "1")
            @PathVariable Long id) {
        try {
            employeeService.deactivateEmployee(id);
            return ResponseEntity.ok().build();
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    @PreAuthorize("hasAuthority('admin')")
    @Operation(summary = "Activate an employee", description = "Activates an employee by their ID.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Employee activated successfully"),
            @ApiResponse(responseCode = "404", description = "Employee not found")
    })
    @PatchMapping("/{id}/activate")
    public ResponseEntity<Void> activateEmployee(
            @Parameter(description = "Employee ID", required = true, example = "1")
            @PathVariable Long id) {
        try {
            employeeService.activateEmployee(id);
            return ResponseEntity.ok().build();
        } catch (RuntimeException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
        }
    }

    @PreAuthorize("hasAuthority('admin')")
    @Operation(summary = "Create an employee", description = "Creates an employee.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "201", description = "Employee created successfully"),
            @ApiResponse(responseCode = "400", description = "Input values in wrong format"),
            @ApiResponse(responseCode = "500", description = "Employee creation failed")
    })
    @PostMapping
    public ResponseEntity<?> createEmployee(
            @RequestBody @Valid CreateEmployeeDto createEmployeeDTO,
            BindingResult result
    ){
        if (result.hasErrors()) {
            List<String> errors = result.getAllErrors().stream()
                    .map(ObjectError::getDefaultMessage)
                    .collect(Collectors.toList());
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new ValidationErrorMessageDto(errors));
        }

        try {
            EmployeeDto employeeDto = employeeService.createEmployee(createEmployeeDTO);
            return ResponseEntity.status(HttpStatus.CREATED).body(employeeDto);
        }
        catch (UserAlreadyExistsException | EmailAlreadyExistsException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new ErrorMessageDto(e.getMessage()));
        }
        catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }

    @PreAuthorize("hasAuthority('admin')")
    @Operation(summary = "Update an employee", description = "Updates an employee.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Employee updated successfully"),
            @ApiResponse(responseCode = "400", description = "Input values in wrong format"),
            @ApiResponse(responseCode = "404", description = "Employee not found"),
            @ApiResponse(responseCode = "500", description = "Employee update failed")
    })
    @PutMapping("/{id}")
    public ResponseEntity<?> updateEmployee(
            @Parameter(description = "Employee ID", required = true, example = "1") @PathVariable Long id,
            @RequestBody @Valid UpdateEmployeeDto updateEmployeeDTO,
            BindingResult result
    ) {
        if (result.hasErrors()) {
            List<String> errors = result.getAllErrors().stream()
                    .map(ObjectError::getDefaultMessage)
                    .collect(Collectors.toList());
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new ValidationErrorMessageDto(errors));
        }
        try {
            EmployeeDto employeeDto = employeeService.updateEmployee(id, updateEmployeeDTO);
            return ResponseEntity.status(HttpStatus.OK).body(employeeDto);

        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }
}

File: user-service-kotlin/src/main/kotlin/pack/userservicekotlin/controller/EmployeeController.kt
package pack.userservicekotlin.controller

import jakarta.validation.Valid
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.*
import pack.userservicekotlin.arrow.EmployeeServiceError
import pack.userservicekotlin.domain.dto.employee.CreateEmployeeDto
import pack.userservicekotlin.domain.dto.employee.EmployeeResponseDto
import pack.userservicekotlin.domain.dto.employee.UpdateEmployeeDto
import pack.userservicekotlin.service.EmployeeService
import pack.userservicekotlin.swagger.EmployeeApiDoc

@RestController
@RequestMapping("/api/admin/employees")
class EmployeeController(
    private val employeeService: EmployeeService,
) : EmployeeApiDoc {
    @PreAuthorize("hasRole('EMPLOYEE')")
    @GetMapping("/{id}")
    override fun getEmployeeById(
        @PathVariable id: Long,
    ): ResponseEntity<Any> =
        employeeService.findById(id).fold(
            ifLeft = {
                when (it) {
                    EmployeeServiceError.NotFound -> ResponseEntity.status(HttpStatus.NOT_FOUND).body("Employee not found")
                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { ResponseEntity.ok(it) },
        )

    @PreAuthorize("hasRole('EMPLOYEE')")
    @GetMapping
    override fun getAllEmployees(
        @RequestParam(required = false) firstName: String?,
        @RequestParam(required = false)lastName: String?,
        @RequestParam(required = false) email: String?,
        @RequestParam(required = false) position: String?,
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
    ): ResponseEntity<Page<EmployeeResponseDto>> {
        val pageRequest = PageRequest.of(page, size)
        return ResponseEntity.ok(employeeService.findAll(firstName, lastName, email, position, pageRequest))
    }

    @PreAuthorize("hasRole('ADMIN')")
    @PostMapping
    override fun createEmployee(
        @RequestBody @Valid createEmployeeDTO: CreateEmployeeDto,
    ): ResponseEntity<Any> =
        employeeService.createEmployee(createEmployeeDTO).fold(
            ifLeft = {
                when (it) {
                    EmployeeServiceError.RoleNotFound -> ResponseEntity.status(HttpStatus.NOT_FOUND).body("Role not found")
                    EmployeeServiceError.EmailAlreadyExists -> ResponseEntity.status(HttpStatus.CONFLICT).body("Email already exists")
                    EmployeeServiceError.JmbgAlreadyExists -> ResponseEntity.status(HttpStatus.CONFLICT).body("Jmbg already exists")
                    EmployeeServiceError.UsernameAlreadyExists -> ResponseEntity.status(HttpStatus.CONFLICT).body("Username already exists")

                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { ResponseEntity.status(HttpStatus.CREATED).body(it) },
        )

    @PreAuthorize("hasRole('ADMIN')")
    @PutMapping("/{id}")
    override fun updateEmployee(
        @PathVariable id: Long,
        @RequestBody @Valid updateEmployeeDTO: UpdateEmployeeDto,
    ): ResponseEntity<Any> =
        employeeService.updateEmployee(id, updateEmployeeDTO).fold(
            ifLeft = {
                when (it) {
                    EmployeeServiceError.NotFound -> ResponseEntity.status(HttpStatus.NOT_FOUND).body("Employee not found")
                    EmployeeServiceError.RoleNotFound -> ResponseEntity.status(HttpStatus.NOT_FOUND).body("Role not found")
                    EmployeeServiceError.LimitNotFound -> ResponseEntity.status(HttpStatus.NOT_FOUND).body("Limit not found")
                    else -> ResponseEntity.internalServerError().build()
                }
            },
            ifRight = { ResponseEntity.ok(it) },
        )

    @PreAuthorize("hasRole('ADMIN')")
    @DeleteMapping("/{id}")
    override fun deleteEmployee(
        @PathVariable id: Long,
    ): ResponseEntity<Void> =
        employeeService.deleteEmployee(id).fold(
            ifLeft = { ResponseEntity.status(HttpStatus.NOT_FOUND).build() },
            ifRight = { ResponseEntity.ok().build() },
        )

    @PreAuthorize("hasRole('ADMIN')")
    @PatchMapping("/{id}/deactivate")
    override fun deactivateEmployee(
        @PathVariable id: Long,
    ): ResponseEntity<Void> =
        employeeService.deactivateEmployee(id).fold(
            ifLeft = { ResponseEntity.status(HttpStatus.NOT_FOUND).build() },
            ifRight = { ResponseEntity.ok().build() },
        )

    @PreAuthorize("hasRole('ADMIN')")
    @PatchMapping("/{id}/activate")
    override fun activateEmployee(
        @PathVariable id: Long,
    ): ResponseEntity<Void> =
        employeeService.activateEmployee(id).fold(
            ifLeft = { ResponseEntity.status(HttpStatus.NOT_FOUND).build() },
            ifRight = { ResponseEntity.ok().build() },
        )

    @PreAuthorize("hasRole('EMPLOYEE')")
    @GetMapping("/me")
    override fun getCurrentEmployee(): ResponseEntity<Any> {
        val auth =
            SecurityContextHolder.getContext().authentication
                ?: return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build()

        val email = auth.name
        return employeeService.findByEmail(email).fold(
            ifLeft = { ResponseEntity.status(HttpStatus.NOT_FOUND).body("Employee not found") },
            ifRight = { ResponseEntity.ok(it) },
        )
    }
}
